### DISCUSSION TITLE
3D Raster - Issue or wrong definition?

### DISCUSSION THREAD

**Discussion by hem-28** [kip-hart/MicroStructPy#69] [Q&A] [answered]
Hi,

I believe one could generate a 3D Raster with the latest version? I tried with the following input, but I end up with an index error 'IndexError: index 0 is out of bounds for axis 0 with size 0'
Any thoughts on a solution? Thanks in advance.

    <material>
        <fraction> 1 </fraction>
        <shape> ellipsoid </shape>
        <ratio_ab > 1.34 </ratio_ab >
        <ratio_ac > 2.00 </ratio_ac >
        <ratio_bc > 1.00 </ratio_bc >
        <size>
            <dist_type> norm </dist_type>
            <loc> 10 </loc>
            <scale> 3 </scale>
        </size>
        <position>  
            <dist_type> uniform </dist_type>
            <loc> 0.0 </loc>
            <scale> 100.0 </scale>
        </position>
        <position> 
            <dist_type> uniform </dist_type>
            <loc> 0.0 </loc>
            <scale> 100 </scale>
        </position>
        <position> 
            <dist_type> norm </dist_type>
            <loc> 50.0 </loc>
            <scale> 5.0 </scale>
        </position>
        <name> Mat-A </name>
    </material>

    <material>
        <fraction> 4 </fraction>
        <shape> ellipsoid </shape>
        <ratio_ab > 1.34 </ratio_ab >
        <ratio_ac > 2.00 </ratio_ac >
        <ratio_bc > 1.00 </ratio_bc >
        <size>
            <dist_type> norm </dist_type>
            <loc> 15 </loc>
            <scale> 3 </scale>
        </size>
        <name> Mat-B </name>
    </material>

    <domain>
        <shape> cube </shape>
        <side_length> 100 </side_length>
        <corner> (0, 0, 0) </corner>
    </domain>
    <settings>
        <directory> two_phase_3D </directory>
        <verbose> True </verbose>
        <filetypes>
            <seeds> vtk </seeds>
            <poly> txt </poly>
			<tri> txt </tri>
            <seeds_plot> png </seeds_plot>
            <poly_plot> png </poly_plot>
			<tri_plot> png </tri_plot>
            <verify_plot> png </verify_plot>
        </filetypes>
        <mesher> raster </mesher>
        <seeds_kwargs>
            <linewidths> 0.2 </linewidths>
        </seeds_kwargs>
        <poly_kwargs>
            <linewidths> 0.2 </linewidths>
        </poly_kwargs>
    </settings>

**Comment by kip-hart** [comment]
Hi @hem-28, thanks for sharing your input file. I can reproduce the error on my computer and I'll find a solution this week. I appreciate your patience.

**Comment by kip-hart** [comment, marked as answer]
Hi @hem-28, I found the issue. The raster mesh generator uses mesh_size to determine the edge length for each pixel/voxel. The default value is infinity, which produces nonsense results internally. When I set mesh_size to 5, I get this result:

![rastermesh](https://user-images.githubusercontent.com/41959581/173716304-ea7a1d60-0fc8-44d4-9c36-aabdc14c20f7.png)

The input file should be as below, with the mesh_size setting indicated. You can set mesh_size to a value you prefer. I considered setting it to 1, but that is too many voxels for my computer. Let me know if you need any further assistance.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<input>
<material>
    <fraction> 1 </fraction>
    <shape> ellipsoid </shape>
    <ratio_ab > 1.34 </ratio_ab >
    <ratio_ac > 2.00 </ratio_ac >
    <ratio_bc > 1.00 </ratio_bc >
    <size>
        <dist_type> norm </dist_type>
        <loc> 10 </loc>
        <scale> 3 </scale>
    </size>
    <position>  <!-- x -->
        <dist_type> uniform </dist_type>
        <loc> 0.0 </loc>
        <scale> 100.0 </scale>
    </position>
    <position> <!-- y -->
        <dist_type> uniform </dist_type>
        <loc> 0.0 </loc>
        <scale> 100 </scale>
    </position>
    <position> <!-- z -->
        <dist_type> norm </dist_type>
        <loc> 50.0 </loc>
        <scale> 5.0 </scale>
    </position>
    <name> Mat-A </name>
</material>

<material>
    <fraction> 4 </fraction>
    <shape> ellipsoid </shape>
    <ratio_ab > 1.34 </ratio_ab >
    <ratio_ac > 2.00 </ratio_ac >
    <ratio_bc > 1.00 </ratio_bc >
    <size>
        <dist_type> norm </dist_type>
        <loc> 15 </loc>
        <scale> 3 </scale>
    </size>
    <name> Mat-B </name>
</material>

<domain>
    <shape> cube </shape>
    <side_length> 100 </side_length>
    <corner> (0, 0, 0) </corner>
</domain>
<settings>
    <directory> two_phase_3D </directory>
    <verbose> True </verbose>
    <filetypes>
        <seeds> vtk </seeds>
        <poly> txt </poly>
		<tri> txt </tri>
        <seeds_plot> png </seeds_plot>
        <poly_plot> png </poly_plot>
		<tri_plot> png </tri_plot>
        <verify_plot> png </verify_plot>
    </filetypes>
    <mesher> raster </mesher>
    <mesh_size> 5 </mesh_size>  <!-- this value sets the edge length of each voxel -->
    <seeds_kwargs>
        <linewidths> 0.2 </linewidths>
    </seeds_kwargs>
    <poly_kwargs>
        <linewidths> 0.2 </linewidths>
    </poly_kwargs>
</settings>
</input>
```

**Comment by hem-28** [comment]
@kip-hart  Thanks for your quick solution. It worked. I tried with a mesh size of 1, that worked too.
![image](https://user-images.githubusercontent.com/89460629/173765174-9d934fe0-8a43-4811-bebf-4bf5851deae1.png)
